### PR TITLE
added two new caches: FileCache, MemoryCache

### PR DESCRIPTION
**Pull request by simedw**
Adds two new caches, FileCache and MemoryCache. 
FileCache is used by default from the CLI. 

```python
from benchllm.chache import FileCache
evaluator = FileCache(StringMatchEvaluator(workers=2), Path("path/to/cache.json"))
```
or 

```python
from benchllm.chache import MemoryCache
evaluator = MemoryCache(StringMatchEvaluator(workers=2))
```

from the cli (same from `bench eval`)

```bash
$ bench run --cache file # default
$ bench run --cache memory
$ bench run --cache none # old behaviour
```

Example cache
```
cat output/cache.json 
{
    "entries": {
        "[\"I was created by V7\", \"I was created by V7.\"]": true,
        "[\"2\", \"2\"]": true,
        "[\"Hello, user!\", \"V7\"]": false,
        "[\"Hello, user!\", \"I was created by V7\"]": false,
        "[\"2\", \"Hello, user!\"]": false,
        "[\"2.0\", \"Hello, user!\"]": false,
        "[\"False\", \"False\"]": true,
        "[\"False\", \"True\"]": false
    },
    "version": "1"
}
```

There is most likely a better way to store the keys, but right now we encode them to json strings, this way strings, dicts, ints, lists are all supported for `expected` and for `output`. Didn't want to shoehorn us into only supporting strings. 

The cli also indicates if the cache was used or not. 

image

### CODE REVIEW

File: benchllm/cli/commands/evaluate.py [hunk 1]
```diff
@@ -17,6 +18,10 @@ def evaluate_predictions(
     load_prediction_files(file_or_dir)
 
     evaluator = get_evaluator(evaluator_name, model, workers)
+    evaluator = add_cache(cache, evaluator, output_dir / ".." / f"cache.json")
```
Comment (andreaazzini) [inline benchllm/cli/commands/evaluate.py:21]: You should be able to use `Path.parent` here, with no hardcoded `".."` value. Also, we don't need an f-string here.

```suggestion
    evaluator = add_cache(cache, evaluator, output_dir.parent / "cache.json")
```

File: benchllm/cache.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        for expected in prediction.test.expected:
+            lookup = self.lookup(expected, prediction.output)
+            # None indicates that nothing was found in the cache
+            # while True and False are both valid cache values
+            if lookup is None:
+                continue
+            self._num_cache_hits += 1
+            if lookup:
+                return Evaluator.Match(prediction=prediction.output, expected=expected)
+            return None
```
Comment (andreaazzini) [inline benchllm/cache.py:42]: I am not sure I get the logic here. If we miss, we want to break for this loop, not return, right? Otherwise, we don't count the miss (line 44).

File: benchllm/cli/commands/run_suite.py [hunk 3]
```diff
@@ -45,6 +47,10 @@ def run_suite(
         return True
 
     evaluator = get_evaluator(evaluator_name, model, workers)
+    evaluator = add_cache(cache, evaluator, output_dir / ".." / f"cache.json")
```
Comment (andreaazzini) [inline benchllm/cli/commands/run_suite.py:50]: Same here
```suggestion
    evaluator = add_cache(cache, evaluator, output_dir.parent / "cache.json")
```

File: benchllm/evaluator.py [hunk 4]
```diff
@@ -5,11 +5,13 @@
 from operator import attrgetter
 from pathlib import Path
 from timeit import default_timer as timer
-from typing import List
+from typing import Any, List, Optional
```
Comment (andreaazzini) [inline benchllm/evaluator.py:8]: I don't think we are using `Any` anywhere
```suggestion
from typing import List, Optional
```

File: benchllm/input_types.py [hunk 5]
```diff
@@ -1,8 +1,10 @@
 import json
-from typing import TypedDict
+from typing import TypedDict, Union
 
 from pydantic import BaseModel
 
+Json = Union[str, bool, list, dict]
```
Comment (andreaazzini) [inline benchllm/input_types.py:6]: I don't think this is the right way of typing a JSON-serializable object/variable. I found a very interesting issue in the Python official repo: https://github.com/python/typing/issues/182.

I really like this approach:
> All major type checkers now support recursive type aliases by default, so this should largely work:

```python
JSON: TypeAlias = dict[str, "JSON"] | list["JSON"] | str | int | float | bool | None
```

What do you think?

**Review by andreaazzini — CHANGES_REQUESTED** [review]
I love these two new Evaluators! I've added a few comments that I think are important to address.
When you have a chance, please take a look and let me know your thoughts on them.

File: benchllm/input_types.py (hunk 5 above)
Comment (simedw) [reply benchllm/input_types.py:6]: Interesting I will check it out, I removed int and float on purpose since it made pydantic parse things in a strange way

```python 
"2" -> 2
["2"] -> ["2"] 
```
which was kind of unexpected and hard to work with.

File: benchllm/cache.py (hunk 2 above)
Comment (simedw) [reply benchllm/cache.py:42]:
```python
for expected in prediction.test.expected:
            lookup = self.lookup(expected, prediction.output)
            # None indicates that nothing was found in the cache
            # while True and False are both valid cache values
            if lookup is None:
                continue
            self._num_cache_hits += 1
            if lookup:
                return Evaluator.Match(prediction=prediction.output, expected=expected)
            return None
```

If lookup is None, then we don't have any entry for that prediction, expected tuple so we continue. 
If we have an entry, it might be a positive entry (they match) and then we can stop searching. 
If we have an entry, it might also be a negative entry (they do not match) and we can also stop searching [this could be debated tbh]

File: benchllm/input_types.py (hunk 5 above)
Comment (simedw) [reply benchllm/input_types.py:6]: https://github.com/pydantic/pydantic/discussions/5779

Seems like Pydantic does not yet support recursive TypeAlias

File: benchllm/cache.py (hunk 2 above)
Comment (simedw) [reply benchllm/cache.py:48]: Have a look now, your comment helped me find a corner case which I also wrote a test case for!

Comment (andreaazzini) [reply benchllm/cache.py:48]: I see, thank you for adding some documentation, it's much clearer now!

**Review by andreaazzini — APPROVED** [review]
I think this looks really good, especially:
- better inline comments as dev-friendly documentation
- split of evaluators into individual modules
- better and larger test coverage

Approved! 👍
